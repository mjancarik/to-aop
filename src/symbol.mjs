export const AOP_PATTERN = Symbol('AopPattern');
export const AOP_HOOKS = Symbol('AopHooks');
export const AOP_STATIC_ALLOW = Symbol('AopStaticAllow');
export const AOP_FILTER_FUNCTION = Symbol('AopFilterFunction');
