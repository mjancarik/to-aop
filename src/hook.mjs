import { AOP_FILTER_FUNCTION } from './symbol.mjs';

export function createHook(name, regular, callback) {
  function hook(meta) {
    return hookFor(meta, regular, callback);
  }

  hook[AOP_FILTER_FUNCTION] = function ({ property, target, prototype }) {
    return (
      (typeof regular === 'string' && property.includes(regular)) ||
      (regular instanceof RegExp && regular.test(property)) ||
      (typeof regular === 'function' &&
        regular({ property, target, name, prototype }))
    );
  };

  return {
    [name]: hook,
  };
}

export function hookFor(meta, regular, callback) {
  if (typeof regular === 'string') {
    if (meta.property.includes(regular)) {
      return callback(meta);
    }

    return null;
  }
  if (regular instanceof RegExp) {
    if (regular.test(meta.property)) {
      return callback(meta);
    }

    return null;
  }

  if (typeof regular === 'function') {
    if (regular(meta)) {
      return callback(meta);
    }

    return null;
  }

  if (Array.isArray(regular)) {
    return regular.map(({ rule, action }) => {
      return hookFor(meta, rule, action || callback);
    });
  }

  throw new TypeError(
    `Invalid rule type ${typeof regular}. Method accept string, regexp, function and array.`,
  );
}

export function hasToRegisterHook(hooks) {
  return (pattern, props) => {
    return hooks.reduce((result, hook) => {
      return (
        result ||
        (pattern[hook] &&
          typeof pattern[hook][AOP_FILTER_FUNCTION] === 'function' &&
          pattern[hook][AOP_FILTER_FUNCTION](props))
      );
    }, false);
  };
}

export const hookName = Object.freeze({
  beforeMethod: 'beforeMethod',
  afterMethod: 'afterMethod',
  aroundMethod: 'aroundMethod',
  beforeGetter: 'beforeGetter',
  afterGetter: 'afterGetter',
  aroundGetter: 'aroundGetter',
  beforeSetter: 'beforeSetter',
  afterSetter: 'afterSetter',
  aroundSetter: 'aroundSetter',
});
