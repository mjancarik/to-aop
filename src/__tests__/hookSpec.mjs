import { expect, jest, describe, beforeEach, it } from '@jest/globals';

import { createHook, hookFor, hookName, hasToRegisterHook } from '../hook.mjs';
import { AOP_FILTER_FUNCTION } from '../symbol.mjs';
import createPattern from './createPattern.mjs';

describe('hook', () => {
  describe('method createHook', () => {
    it('should return specific hook structure', () => {
      const hook = createHook(hookName.beforeMethod, 'setMethod', () => {});

      expect(hook).toMatchSnapshot();
    });

    it('should return true if method match with defined regular', () => {
      const hook = createHook(hookName.beforeMethod, 'setMethod', () => {});

      expect(
        hook[hookName.beforeMethod][AOP_FILTER_FUNCTION]({
          property: 'setMethod',
        }),
      ).toEqual(true);
    });

    it('should return false if method not match with defined regular', () => {
      const hook = createHook(hookName.beforeMethod, 'setMethod', () => {});

      expect(
        hook[hookName.beforeMethod][AOP_FILTER_FUNCTION]({
          property: 'getMethod',
        }),
      ).toEqual(false);
    });
  });

  describe('method hookFor', () => {
    let metaTruthy = null;
    let metaFalsy = null;
    let fn = null;

    beforeEach(() => {
      metaTruthy = {
        property: 'setState',
      };
      metaFalsy = {
        property: 'getState',
      };
      fn = jest.fn();
    });

    it('should call action for filtering rule as string', () => {
      hookFor(metaTruthy, 'setState', fn);

      expect(fn).toHaveBeenCalledWith(metaTruthy);
    });

    it('should call action for filtering rule as regexp', () => {
      hookFor(metaTruthy, /setState/gi, fn);

      expect(fn).toHaveBeenCalledWith(metaTruthy);
    });

    it('should call action for filtering rule as function', () => {
      hookFor(metaTruthy, () => true, fn);

      expect(fn).toHaveBeenCalledWith(metaTruthy);
    });

    it('should call action hook for filtering rule as list', () => {
      hookFor(metaTruthy, [{ rule: 'setState', action: fn }]);

      expect(fn).toHaveBeenCalledWith(metaTruthy);
    });

    it('should throw error for bad structure at filtering rule', () => {
      expect(() => hookFor(metaTruthy, {}, fn)).toThrow();
    });

    it('should not call action', () => {
      hookFor(metaFalsy, () => false, fn);

      expect(fn).not.toHaveBeenCalled();
    });
  });

  describe('method hasToRegisterHook', () => {
    let pattern;
    let hasToRegisterGetterHook;

    beforeEach(() => {
      pattern = createPattern(
        {
          [hookName.beforeGetter]: ({ property }) =>
            property === 'staticGetter',
        },
        {
          [hookName.beforeGetter]: () => {},
        },
      );

      hasToRegisterGetterHook = hasToRegisterHook([hookName.beforeGetter]);
    });

    it('should return true if hook has to be registered', () => {
      expect(
        hasToRegisterGetterHook(pattern, { property: 'staticGetter' }),
      ).toEqual(true);
    });

    it('should return false if hook has not to be registered', () => {
      expect(
        hasToRegisterGetterHook(pattern, { property: 'staticGetter2' }),
      ).toEqual(false);
    });
  });
});
