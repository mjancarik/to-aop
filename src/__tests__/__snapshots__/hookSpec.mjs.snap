// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hook method createHook should return specific hook structure 1`] = `
{
  "beforeMethod": [Function],
}
`;
